orchagent/policerorch.cpp

```diff
@@ -123,119 +142,395 @@ void PolicerOrch::doTask(Consumer &consumer)
+            }
+            else if (field == red_packet_action_field)
+            {
+                attr.id = SAI_POLICER_ATTR_RED_PACKET_ACTION;
+                attr.value.s32 = packet_action_map.at(value);
+            }
+            else if (field == green_packet_action_field)
+            {
+                attr.id = SAI_POLICER_ATTR_GREEN_PACKET_ACTION;
+                attr.value.s32 = packet_action_map.at(value);
+            }
+            else if (field == yellow_packet_action_field)
+            {
+                attr.id = SAI_POLICER_ATTR_YELLOW_PACKET_ACTION;
+                attr.value.s32 = packet_action_map.at(value);
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Unknown policer attribute %s specified",
+                        field.c_str());
+                continue;
+            }
+
+            attrs.push_back(attr);
+        }
+
+        // Create a new policer
+        if (!update)
+        {
+            if (!meter_type || !mode)
+            {
+                SWSS_LOG_ERROR("Failed to create policer %s,\
+                        missing madatory fields", key.c_str());
+                return task_process_status::task_invalid_entry;
+            }
+
+            sai_object_id_t policer_id;
+            sai_status_t status = sai_policer_api->create_policer(
+                    &policer_id, gSwitchId, (uint32_t)attrs.size(), attrs.data());
+            if (status != SAI_STATUS_SUCCESS)
+            {
+                SWSS_LOG_ERROR("Failed to create policer %s, rv:%d",
+                        key.c_str(), status);
+                return task_process_status::task_need_retry;
+            }
+
+            SWSS_LOG_NOTICE("Created policer %s", key.c_str());
+            m_syncdPolicers[key].policerOid = policer_id;
+            m_policerRefCounts[key] = 0;
+        }
+        else
+        {
+            auto policer_id = m_syncdPolicers[key].policerOid;
+
+            // The update operation has limitations that it could only update
+            // the rate and the size accordingly.
+            // SR_TCM: CIR, CBS, PBS
+            // TR_TCM: CIR, CBS, PIR, PBS
+            // STORM_CONTROL: CIR, CBS
+            for (auto & attr: attrs)
+            {
+                if (attr.id != SAI_POLICER_ATTR_CBS &&
+                        attr.id != SAI_POLICER_ATTR_CIR &&
+                        attr.id != SAI_POLICER_ATTR_PBS &&
+                        attr.id != SAI_POLICER_ATTR_PIR)
+                {
+                    continue;
+                }
+
+                sai_status_t status = sai_policer_api->set_policer_attribute(
+                        policer_id, &attr);
+                if (status != SAI_STATUS_SUCCESS)
+                {
+                    SWSS_LOG_ERROR("Failed to update policer %s attribute, rv:%d",
+                            key.c_str(), status);
+                    return task_process_status::task_need_retry;
```

Comment:
The initial idea was to avoid partial update to the policer when one of the attributes fail to get updated. 
But if it is recommended to continue with remaining attributes, then this can be changed to continue instead of return. Please suggest.

Reply:
As long as SAI is capable of handling the partial updates,  IMO, we can continue with the partial updates to SAI, we don't have to retry multiple times.